File: src/Domain/Entities/State.cs
using System.Collections.Generic;

namespace CDACommercial.PoC.Domain.Entities
{
    public class State : Entity
    {
        public string Code { get; private set; }
        public string Name { get; private set; }
        public int TotalCities {get; private set; }
        public int TotalListings {get; private set; }
        public List<City> Cities { get; private set; }
        public State(string code, string name)
        {
            Code = code;
            Name = name ?? "";
            Cities = new List<City>();
            TotalCities = 0;
        }

        public void UpdateStats(int totalCities, int totalListings)
        {
            TotalCities = totalCities;
            TotalListings = totalListings;
        }
    }
}

File: src/Application/ClientApp/src/app/api/types/state.ts
import { City } from "./city";

export class State {
    name: string;
    cities: City[] = [];
    validCities: City[] = [];
    totalCities: number = 0;
    totalListings: number = 0;

    get slug(): string {
        return this.name.toLocaleLowerCase();
    }

    constructor(data: any) {
        if (!data) { return; }
        for (let property in data) {
            if (!data.hasOwnProperty(property)) { continue; }
            switch (property) {
                case 'cities':
                    this.cities = City.fromJSCollection(data[property]);
                    this.validCities = this.cities.filter(m => !!m.code);
                    break;
                default:
                    (<any>this)[property] = data[property];
                    break;
            }
        }
    }

    static fromJS(data: any): State {
        data = typeof data === 'object' ? data : {};
        return new State(data);
    }

    static fromJSCollection(data: any[]): State[] {
        if (!data) return null;
        return data.map(d => State.fromJS(d));
    }
}